src/nvim/diff.c

```diff
@@ -2637,3 +2985,148 @@ linenr_T diff_lnum_win(linenr_T lnum, win_T *wp)
+                         linenr_T    *lnum_new,
+                         long       *count_new)
+{
+  char_u *p;
+  long    f1, l1, f2, l2;
+  int       difftype;
+
+  // The line must be one of three formats:
+  // change: {first}[,{last}]c{first}[,{last}]
+  // append: {first}a{first}[,{last}]
+  // delete: {first}[,{last}]d{first}
+  p = line;
+  f1 = getdigits(&p);
+  if (*p == ',') {
+    p++;
+    l1 = getdigits(&p);
+  } else {
+    l1 = f1;
+  }
+  if (*p != 'a' && *p != 'c' && *p != 'd') {
+    return FAIL;        // invalid diff format
+  }
+  difftype = *p++;
+  f2 = getdigits(&p);
+  if (*p == ',') {
+    p++;
+    l2 = getdigits(&p);
+  } else {
+    l2 = f2;
+  }
+  if (l1 < f1 || l2 < f2) {
+    return FAIL;
+  }
+
+  if (difftype == 'a') {
+    *lnum_orig = f1 + 1;
+    *count_orig = 0;
+  } else {
+    *lnum_orig = f1;
+    *count_orig = l1 - f1 + 1;
+  }
+  if (difftype == 'd') {
+    *lnum_new = f2 + 1;
+    *count_new = 0;
+  } else {
+    *lnum_new = f2;
+    *count_new = l2 - f2 + 1;
+  }
+  return OK;
+}
+
+///
+/// Parses unified diff with zero(!) context lines.
+/// Return FAIL if there is no diff information in "line".
+///
+static int parse_diff_unified(char_u        *line,
+                              linenr_T    *lnum_orig,
+                              long      *count_orig,
+                              linenr_T    *lnum_new,
+                              long      *count_new)
+{
+  char_u *p;
+  long    oldline, oldcount, newline, newcount;
+
+  // Parse unified diff hunk header:
+  // @@ -oldline,oldcount +newline,newcount @@
+  p = line;
+  if (*p++ == '@' && *p++ == '@' && *p++ == ' ' && *p++ == '-') {
+    oldline = getdigits(&p);
+    if (*p == ',') {
+      p++;
+      oldcount = getdigits(&p);
+    } else {
+      oldcount = 1;
+    }
+    if (*p++ == ' ' && *p++ == '+') {
+      newline = getdigits(&p);
+      if (*p == ',') {
+        p++;
+        newcount = getdigits(&p);
+      } else {
+        newcount = 1;
+      }
+    } else {
+      return FAIL;  // invalid diff format
+    }
+
+    if (oldcount == 0) {
+      oldline += 1;
+    }
+    if (newcount == 0) {
+      newline += 1;
+    }
+    if (newline == 0) {
+      newline = 1;
+    }
+
+    *lnum_orig = oldline;
+    *count_orig = oldcount;
+    *lnum_new = newline;
+    *count_new = newcount;
+
+    return OK;
+  }
+
+  return FAIL;
+}
+
+///
+/// Callback function for the xdl_diff() function.
+/// Stores the diff output in a grow array.
+///
+static int xdiff_out(void *priv, mmbuffer_t *mb, int nbuf)
+{
+  diffout_T *dout = (diffout_T *)priv;
+  char_u    *p;
+
+  // The header line always comes by itself, text lines in at least two
+  // parts.  We drop the text part.
+  if (nbuf > 1) {
+    return 0;
+  }
+
+  // sanity check
+  if (STRNCMP(mb[0].ptr, "@@ ", 3)  != 0) {
+    return 0;
+  }
+
+  ga_grow(&dout->dout_ga, 1);
+
+  p = vim_strnsave((char_u *)mb[0].ptr, mb[0].size);
+  if (p == NULL) {
```

Comment:
can never be NULL. coverity or some other analyzer will warn about this, so let's remove this check now.

Reply:
Done